helm-app-operator/pkg/helm/helm.go

```diff
@@ -271,32 +278,116 @@ func (c installer) UninstallRelease(r *unstructured.Unstructured) (*unstructured
 		return r, nil
 	}
 
-	tiller := tillerRendererForCR(r, c.storageBackend, c.tillerKubeClient)
+	tiller := c.tillerRendererForCR(r)
 	_, err = tiller.UninstallRelease(context.TODO(), &services.UninstallReleaseRequest{
 		Name:  releaseName(r),
 		Purge: true,
 	})
 	if err != nil {
 		return r, err
 	}
-
+	logrus.Infof("Uninstalled release for %s release=%s", ResourceString(r), releaseName(r))
 	return r, nil
 }
 
-func (c installer) syncReleaseStatus(status v1alpha1.HelmAppStatus) {
+// ResourceString returns a human friendly string for the custom resource
+func ResourceString(r *unstructured.Unstructured) string {
+	return fmt.Sprintf("apiVersion=%s kind=%s name=%s/%s", r.GetAPIVersion(), r.GetKind(), r.GetNamespace(), r.GetName())
+}
+
+func (c installer) installRelease(r *unstructured.Unstructured, tiller *tiller.ReleaseServer, chart *cpb.Chart, config *cpb.Config) (*release.Release, error) {
+	installReq := &services.InstallReleaseRequest{
+		Namespace: r.GetNamespace(),
+		Name:      releaseName(r),
+		Chart:     chart,
+		Values:    config,
+	}
+
+	releaseResponse, err := tiller.InstallRelease(context.TODO(), installReq)
+	if err != nil {
+		return nil, err
+	}
+	return releaseResponse.GetRelease(), nil
+}
+
+func (c installer) updateRelease(r *unstructured.Unstructured, tiller *tiller.ReleaseServer, chart *cpb.Chart, config *cpb.Config) (*release.Release, error) {
+	updateReq := &services.UpdateReleaseRequest{
+		Name:   releaseName(r),
+		Chart:  chart,
+		Values: config,
+	}
+
+	releaseResponse, err := tiller.UpdateRelease(context.TODO(), updateReq)
+	if err != nil {
+		return nil, err
+	}
+	return releaseResponse.GetRelease(), nil
+}
+
+func (c installer) reconcileRelease(r *unstructured.Unstructured, expectedManifest string) error {
+	expectedInfos, err := c.tillerKubeClient.BuildUnstructured(r.GetNamespace(), bytes.NewBufferString(expectedManifest))
+	if err != nil {
+		return err
+	}
+	return expectedInfos.Visit(func(expected *resource.Info, err error) error {
+		if err != nil {
+			return err
+		}
+		helper := resource.NewHelper(expected.Client, expected.Mapping)
+		_, err = helper.Create(expected.Namespace, true, expected.Object)
```

Comment:
This is interesting. I've never come across the cli-runtime or [Info](https://github.com/kubernetes/cli-runtime/blob/79bf4e0b64544d8c490247abae089bea572ddae6/pkg/genericclioptions/resource/visitor.go#L60-L62) before. Quite convenient that the tiller kubeclient can provide that for us.